File: Alkitab/src/yuku/alkitab/base/ac/AboutActivity.java
package yuku.alkitab.base.ac;

import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.text.SpannableStringBuilder;
import android.text.method.LinkMovementMethod;
import android.text.style.URLSpan;
import android.util.DisplayMetrics;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import yuku.afw.V;
import yuku.alkitab.base.App;
import yuku.alkitab.base.U;
import yuku.alkitab.base.ac.base.BaseActivity;
import yuku.alkitab.debug.R;

public class AboutActivity extends BaseActivity {
	public static final String TAG = AboutActivity.class.getSimpleName();

	View root;
	TextView tVersion;
	TextView tTranslators;
	ImageView imgLogo;

	View bHelp;
	View bDonation;
	View bSuggest;

	@Override protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_about);
		setTitle(String.format("%s %s", getString(R.string.app_name), App.getVersionName()));
		getSupportActionBar().setSubtitle(String.format("%s %s", App.getVersionCode(), getString(R.string.last_commit_hash)));

		root = V.get(this, R.id.root);
		tVersion = V.get(this, R.id.tVersion);
		tTranslators = V.get(this, R.id.tTranslators);
		imgLogo = V.get(this, R.id.imgLogo);
		bHelp = V.get(this, R.id.bHelp);
		bDonation = V.get(this, R.id.bDonation);
		bSuggest = V.get(this, R.id.bSuggest);

		Drawable logoDrawable;
		if (Build.VERSION.SDK_INT >= 15) {
			logoDrawable = getResources().getDrawableForDensity(R.drawable.ic_launcher, DisplayMetrics.DENSITY_XXXHIGH);
		} else {
			logoDrawable = getResources().getDrawable(R.drawable.ic_launcher);
		}
		imgLogo.setImageDrawable(logoDrawable);

		tVersion.setText(getString(R.string.about_version_name, App.getVersionName()));

		bHelp.setOnClickListener(bHelp_click);
		bDonation.setOnClickListener(bDonation_click);
		bSuggest.setOnClickListener(bSuggest_click);

		String[] translators = getResources().getStringArray(R.array.translators_list);
		SpannableStringBuilder sb = new SpannableStringBuilder();
		for (String translator: translators) {
			int open = translator.indexOf('[');
			int close = translator.indexOf(']');
			if (open != -1 && close > open) {
				sb.append(translator.substring(0, open));
				int sb_len = sb.length();
				sb.append(translator.substring(close + 1));
				sb.setSpan(new URLSpan(translator.substring(open + 1, close)), sb_len, sb.length(), 0);
			} else {
				sb.append(translator);
			}
			sb.append('\n');
		}
		tTranslators.setText(sb);
		tTranslators.setMovementMethod(LinkMovementMethod.getInstance());

		root.setOnTouchListener(root_touch);
	}

	View.OnTouchListener root_touch = new View.OnTouchListener() {
		@Override
		public boolean onTouch(final View v, final MotionEvent event) {
			if (event.getPointerCount() >= 4) {
				getWindow().setBackgroundDrawable(new GradientDrawable(GradientDrawable.Orientation.BR_TL, new int[] {0xffaaffaa, 0xffaaffff, 0xffaaaaff, 0xffffaaff, 0xffffaaaa, 0xffffffaa}));
			}

			return false;
		}
	};

	View.OnClickListener bHelp_click = new View.OnClickListener() {
		@Override
		public void onClick(final View v) {
			String page;
			if (U.equals("in", getResources().getConfiguration().locale.getLanguage())) {
				page = "help/html-in/index.html";
			} else {
				page = "help/html-en/index.html";
			}

			startActivity(HelpActivity.createIntent(page, false, null, null));
		}
	};

	View.OnClickListener bDonation_click = new View.OnClickListener() {
		@Override
		public void onClick(final View v) {
			String donation_url = getString(R.string.alamat_donasi);
			Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(donation_url));
			startActivity(HelpActivity.createIntent("help/donation.html", true, getString(R.string.send_donation_confirmation), intent));
		}
	};

	View.OnClickListener bSuggest_click = new View.OnClickListener() {
		@Override
		public void onClick(final View v) {
			startActivity(com.example.android.wizardpager.MainActivity.createIntent(App.context));
		}
	};
}

File: Alkitab/src/main/java/yuku/alkitab/base/ac/AboutActivity.kt
package yuku.alkitab.base.ac

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.text.method.LinkMovementMethod
import android.util.DisplayMetrics
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.res.ResourcesCompat
import androidx.core.graphics.ColorUtils
import androidx.core.net.toUri
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean
import yuku.alkitab.base.App
import yuku.alkitab.base.ac.base.BaseActivity
import yuku.alkitab.debug.BuildConfig
import yuku.alkitab.debug.R

class AboutActivity : BaseActivity() {
    private lateinit var root: View
    private lateinit var tVersion: TextView
    private lateinit var tBuild: TextView
    private lateinit var imgLogo: ImageView
    private lateinit var tAboutTextDesc: TextView
    private lateinit var bHelp: View
    private lateinit var bMaterialSources: View
    private lateinit var bCredits: View

    private val backgroundAnimationStarted = AtomicBoolean(false)
    private var baseHue = 0
    private val hsl = FloatArray(3)
    private val colors = IntArray(6)

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_about)

        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        root = findViewById(R.id.root)
        tVersion = findViewById(R.id.tVersion)
        tBuild = findViewById(R.id.tBuild)
        imgLogo = findViewById(R.id.imgLogo)
        tAboutTextDesc = findViewById(R.id.tAboutTextDesc)

        bHelp = findViewById(R.id.bHelp)
        bHelp.setOnClickListener {
            startActivity(Intent(Intent.ACTION_VIEW, "https://alkitab.app/guide?utm_source=app&utm_medium=button&utm_campaign=help".toUri()))
        }

        bMaterialSources = findViewById(R.id.bMaterialSources)
        bMaterialSources.setOnClickListener {
            startActivity(HelpActivity.createIntent("help/material_sources.html", getString(R.string.about_material_sources)))
        }

        bCredits = findViewById(R.id.bCredits)
        bCredits.setOnClickListener {
            startActivity(HelpActivity.createIntent("help/credits.html", getString(R.string.about_credits)))
        }

        imgLogo.setImageDrawable(ResourcesCompat.getDrawableForDensity(resources, R.mipmap.ic_launcher, DisplayMetrics.DENSITY_XXXHIGH, null))
        imgLogo.setOnTouchListener { v: View, event: MotionEvent ->
            if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                val x = event.x
                val y = event.y
                if (x >= v.width / 2 - 4 && x <= v.width / 2 + 4) {
                    if (y >= v.height * 3 / 10 - 4 && y <= v.height * 3 / 10 + 4) {
                        showSecretDialog()
                    }
                }
            }
            false
        }

        tAboutTextDesc.movementMethod = LinkMovementMethod.getInstance()
        tVersion.text = getString(R.string.about_version_name, App.getVersionName())
        tBuild.text = String.format(Locale.US, "%s %s", App.getVersionCode(), BuildConfig.LAST_COMMIT_HASH)
        root.setOnTouchListener { _, event ->
            if (event.pointerCount == 4) {
                startBackgroundAnimation()
            } else if (event.pointerCount == 5 && event.actionMasked == MotionEvent.ACTION_POINTER_DOWN) {
                showSecretDialog()
            }
            false
        }
    }

    private fun startBackgroundAnimation() {
        if (!backgroundAnimationStarted.compareAndSet(false, true)) {
            return
        }

        object : Handler(Looper.getMainLooper()) {
            override fun handleMessage(msg: Message) {
                if (isFinishing) return  // don't leak
                val baseColor = -0x660067
                ColorUtils.colorToHSL(baseColor, hsl)
                for (i in colors.indices) {
                    hsl[0] = ((baseHue + i * 60) % 360).toFloat()
                    colors[i] = ColorUtils.HSLToColor(hsl)
                }
                window.setBackgroundDrawable(GradientDrawable(GradientDrawable.Orientation.BR_TL, colors))
                baseHue += 2
                sendEmptyMessageDelayed(0, 16)
            }
        }.sendEmptyMessage(0)
    }

    private fun showSecretDialog() {
        MaterialAlertDialogBuilder(this)
            .setItems(arrayOf("Secret settings", "Crash me")) { _, index ->
                when (index) {
                    0 -> startActivity(SecretSettingsActivity.createIntent())
                    1 -> throw RuntimeException("Dummy exception from secret dialog.")
                }
            }
            .show()
    }

    companion object {
        @JvmStatic
        fun createIntent(): Intent {
            return Intent(App.context, AboutActivity::class.java)
        }
    }
}